### PR TITLE
Delft 3D Timestep to Seconds Function

### PR DESCRIPTION
**Pull request by browniea**
This PR outputs the time based on the time step in function `convert_time`, `d3d. get_layer_data` and `d3d.get_all_data_points`. A function to get all timestamps 'get_all_timestamps' will return an array of all the timestamps for a data set.

Additions: 
-`get_all_timestamps` : inputs netcdf4 dataframe and return as all the time stamps for that data set 
-`convert_time` : converts a `time_index` to a `time_stamp`  in seconds or vice versa

Changes in 'get_layer_data` 
- 'cords_to_layers' dictionary 
- depth output `z` :` z=[bottom_depth*Layer_percentages[layer_index]]-water_level`
- updated interpolation for bottom_depth and water_level in wdim coordinate system 

Changes in 'get_all_data_points` 
- 'cords_to_layers' dictionary 
- depth `z` is pulled from 'get_layer_data`  and takes into account water level

### CODE REVIEW

File: mhkit/river/io/d3d.py [hunk 1]
```diff
@@ -6,6 +6,86 @@
...
+       A netCDF object that contains spatial data, e.g. velocity or shear stess.
+       These can be generated by running a Delft3D model.  
+
+    Returns
+    -------
+      time_stamps: array
+        An array of positive interger or float the repesents the amount of time in seconds 
+        that simulation has been running
+
+    '''
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:27]: Make sure this is formatted correctly: Spelling, capitalization, new lines etc.

File: mhkit/river/io/d3d.py [hunk 2]
```diff
@@ -6,6 +6,86 @@
...
+    retuns all of the time stamps and their associated time index
+    Parameters
+    ----------
+    data: netcdf4 object 
+       A netCDF object that contains spatial data, e.g. velocity or shear stess.
+       These can be generated by running a Delft3D model.  
+
+    Returns
+    -------
+      time_stamps: array
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:23]: this needs to be one word everywhere `timestamps`

File: mhkit/river/io/d3d.py [hunk 3]
```diff
@@ -6,6 +6,86 @@
 import numpy as np
 import pandas as pd
 import netCDF4
+import warnings
+
+
+def get_all_timestamps (data):
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:12]: I don't think what this returns is truly a timestamp. I think its really "seconds passed". I think we should call this something like `get_all_time` or maybe just `get_time`, or `get_seconds`. Then we need to work that into the return value as well. So maybe its now called 'time' or `seconds`.

File: mhkit/river/io/d3d.py [hunk 4]
```diff
@@ -6,6 +6,86 @@
...
+    -------
+      time_stamps: array
+        An array of positive interger or float the repesents the amount of time in seconds 
+        that simulation has been running
+
+    '''
+    assert type(data)== netCDF4._netCDF4.Dataset, 'data must be netCDF4 object'
+        
+    max_time= data['time'].shape
+    max_time= int(''.join(map(str, max_time)))
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:31]: What about we replace these 2 lines with 
`max_time= data['time'].shape[0]`

File: mhkit/river/io/d3d.py [hunk 5]
```diff
@@ -6,6 +6,86 @@
...
+    time_indexes= np.linspace(0, max_time-1, num=max_time)
+        
+    time_stamps= []        
+    for i in time_indexes:
+        time= np.ma.getdata(data.variables['time'][i], False)
+        time_stamps= np.append(time_stamps, time)
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:37]: Isn't this entire function just doing this:
`np.ma.getdata(data.variables['time'][:], False)`

File: mhkit/river/io/d3d.py [hunk 6]
```diff
@@ -6,6 +6,86 @@
...
+        that simulation has been running
+
+    Returns
+    -------
+    QoI: int, float
+        The quantity of interest is the unknown value either the time_index or
+        the time_stamp. The time_index is a positive integer starting form 0 
+        and incrementing until in simulation is complete. The time_stamp is the 
+        seconds corresponding to those increments.
+
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:65]: no space

File: mhkit/river/io/d3d.py [hunk 7]
```diff
@@ -6,6 +6,86 @@
...
+    Returns
+    -------
+    QoI: int, float
+        The quantity of interest is the unknown value either the time_index or
+        the time_stamp. The time_index is a positive integer starting form 0 
+        and incrementing until in simulation is complete. The time_stamp is the 
+        seconds corresponding to those increments.
+ 
+    '''
+    assert type(data)== netCDF4._netCDF4.Dataset, 'data must be netCDF4 object'
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:67]: add space above

File: mhkit/river/io/d3d.py [hunk 8]
```diff
@@ -40,23 +120,87 @@ def get_layer_data(data, variable, layer_index= -1 , time_index=-1):
     assert variable in data.variables.keys(), 'variable not recognized'
     coords = str(data.variables[variable].coordinates).split()
     var=data.variables[variable][:]
-    max_time_index= len(var)
+    max_time= data['time'].shape
+    max_time_index= int(''.join(map(str, max_time)))-1 # to account for zero index
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:124]: What about: 
`max_time= data['time'].shape[0]`

File: mhkit/river/io/d3d.py [hunk 9]
```diff
@@ -40,23 +120,87 @@ def get_layer_data(data, variable, layer_index= -1 , time_index=-1):
...
+        max_layer= len(var[0][0])
+        
+        assert layer_index <= max_layer,( f'layer_index must be less than the max'
+                                     + 'layer {max_layer}')
+        assert layer_index >= -1, 'layer_index must be greater than or equal to -1'
+        v= np.ma.getdata(var[time_index,:,layer_index], False)
+        dimensions= 3
+    
+    else: 
+        assert type(var[0][0])== np.float64, 'data not reconized'
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:142]: spelling

**Review by ssolson — COMMENTED** [review]
Hey Emily take a look at these comments and let me know what you think

File: mhkit/river/io/d3d.py (hunk 5 above)
Comment (browniea) [reply mhkit/river/io/d3d.py:37]: Yes it is. Do you think it's worth having a whole function for the one line of code?

File: mhkit/river/io/d3d.py (hunk 3 above)
Comment (ssolson) [reply mhkit/river/io/d3d.py:12]: as discussed 

function `get_all_time`
returns `seconds_passed`

File: mhkit/river/io/d3d.py [hunk 10]
```diff
@@ -6,6 +6,79 @@
...
+    assert type(data)== netCDF4._netCDF4.Dataset, 'data must be netCDF4 object'
+        
+
+    seconds_run = np.ma.getdata(data.variables['time'][:], False)
+
+    return seconds_run
+
+def convert_time (data, time_index=None, seconds_run=None):
+    '''
+    Output the seconds_run for a time_index and vice versa. The output captains
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:37]: contains

File: mhkit/river/io/d3d.py [hunk 11]
```diff
@@ -6,6 +6,79 @@
...
+    ----------
+    data: netcdf4 object 
+       A netCDF object that contains spatial data, e.g. velocity or shear stress.
+       These can be generated by running a Delft3D model.  
+    time_index: int 
+        A positive integer to pull the time step from the dataset. 0 being closest
+        to time 0. Default is last time step -1.
+    seconds_run: int, float
+        a positive integer or float the represents the amount of time in seconds 
+        that simulation has been running
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:50]: that the simulation

**Review by ssolson — COMMENTED** [review]
Hey, Emily could you take a look at these comments? Let's do a general review of all docstrings to ensure all the grammar and spelling are clean.

File: mhkit/river/io/d3d.py [hunk 12]
```diff
@@ -40,23 +113,86 @@ def get_layer_data(data, variable, layer_index= -1 , time_index=-1):
...
+    depth=np.append(depth, z)
+
+    time= np.ma.getdata(data.variables['time'][time_index], False)*np.ones(len(x))
+
+    layer= np.array([ [x_i, y_i, z_i, v_i, t_i] for x_i, y_i, z_i, v_i, t_i in
+                     zip(x, y, depth, v, time)]) 
+    layer_data = pd.DataFrame(layer, columns=['x', 'y', 'z','v', 'time'])
+
+
     return layer_data
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:198]: In the example, let's add example input code as the user would input it

File: mhkit/river/io/d3d.py [hunk 13]
```diff
@@ -271,20 +414,26 @@ def get_all_data_points(data, variable, time_index= -1):
...
+    layer_dim =  str(data.variables[variable].coordinates)
     
     try:    
-        cord_sys= cords_to_layers[layer_dim]
+        cord_sys= cords_to_layers[layer_dim]['coords']
+
     except: 
         raise Exception('coordinates not recognized')
     else: 
         Layer_percentages= np.ma.getdata(cord_sys, False)
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:431]: Let's rename to lower case
layer_percentage

**Review by ssolson — COMMENTED** [review]
d3d.py L256
I would use the try and except to define N.

You currently have
```
        try:
            len(directions[i]['values'])
        except:
            directions[i]['values'] = np.array([directions[i]['values']])  
            
        N= len(directions[i]['values'])
```
and make it this:
```
        try:
            N=len(directions[i]['values'])
        except:
            directions[i]['values'] = np.array([directions[i]['values']])  
            N= len(directions[i]['values'])
```

**Comment by browniea** [comment]
Changing the d3d.turbulent_intensity function to output TI in a percentage rather than a fraction 

`TI_data['turbulent_intensity']= np.sqrt(2/3*TI_data['turkin1'])/u_mag * 100`

File: mhkit/river/io/d3d.py [hunk 14]
```diff
@@ -259,125 +398,67 @@ def get_all_data_points(data, variable, time_index= -1):
...
     Parameters
     ----------
-    data: netcdf4 object 
-       A netCDF object that contains spatial data such as velocity or shear stess.
-       These can be generated by running a Delft3D model.  
+    data: Netcdf4 object 
+       A NetCDF4 object that contains spatial data, e.g. velocity or shear
+       stress, generated by running a Delft3D model.  
     points: string, DataFrame  
         Point to interpoate data onto.
```
[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/river/io/d3d.py:463]: L466: coordinates

[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/river/io/d3d.py:463]: Can you please review this and all doc strings for grammatical errors?

[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/river/io/d3d.py:463]: L471: 

    intermediate_values: boolean
        If false x,y,z, and TI values are returned. If True then component velocities are also returned (ucx, uxy, uxz). Deafult  False.

[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/river/io/d3d.py:463]: L479: This explination and deisplayed data structure can be simplified.

File: mhkit/river/io/d3d.py [hunk 15]
```diff
@@ -461,7 +542,7 @@ def turbulent_intensity(data, points='cells', time_index= -1,
     TI_data.loc[zero_ind,'turkin1']=np.zeros(len(zero_ind))
```
Comment (ssolson) [inline mhkit/river/io/d3d.py:542]: L498: Additionally we should check that the provided time index is within the range of the dataset

**Review by ssolson — COMMENTED** [review]
Hey Emily please check out the couple of comments below.

File: mhkit/river/io/d3d.py [hunk 16]
```diff
@@ -259,7 +402,7 @@ def get_all_data_points(data, variable, time_index= -1):
     Returns
     -------
     all_data: DataFrame 
-        Dataframe with columns x, y, z, and variable. 
+        Dataframe with columns x, y, z, variable, and time.
```
[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/river/io/d3d.py:398]: Time index still states it must be positive.

**Review by ssolson — COMMENTED** [review]
Hey, Emily please see a couple of comments here.

File: mhkit/river/io/d3d.py [hunk 17]
```diff
@@ -1,22 +1,94 @@
...
+    specify 'time_index' or 'seconds_run'.
+
+    Parameters
+    ----------
+    data: NetCDF4 object 
+       A NetCDF4 object that contains spatial data, e.g. velocity or shear
+       stress, generated by running a Delft3D model.  
+    time_index: int 
+        A positive integer to pull the time index from the dataset. 0 being closest
+        to time 0. Default is last time index -1.
```
[the remark this answers is not in the document]
Comment (browniea) [reply mhkit/river/io/d3d.py:48]: The function runs with negative indices, should I add more about indexing in d3d.py. I think it's bested explained with an example as it is in Delft3D_example.

File: mhkit/river/io/d3d.py [hunk 18]
```diff
@@ -30,33 +102,94 @@ def get_layer_data(data, variable, layer_index= -1 , time_index=-1):
     Returns
     -------
     layer_data: DataFrame
-        DataFrame with columns of "x" and "y" location on specified layer and 
-        the variable values "v".
+        DataFrame with columns of "x", "y", and "z" location on specified layer,
```
[the remark this answers is not in the document]
Comment (browniea) [reply mhkit/river/io/d3d.py:108]: I defined it as: the value "z" is the depth from the zero-water level at the surface. I did create an issue [#172](https://github.com/MHKiT-Software/MHKiT-Python/issues/172) to change this.

**Pull request merged**